### DISCUSSION TITLE
🐛 Bug Report: Function Error operation timed out

### DISCUSSION THREAD

**Discussion by elio1fiore** [appwrite/appwrite#5058] [Q&A] [answered]
### 👟 Reproduction steps

After setting the dart 2.17 runtime and after initializing the function and writing (basic function to detect the problem):
```dart
Future<void> start(final req, final res) async {
  final client = Client();

  if (req.variables['APPWRITE_FUNCTION_ENDPOINT'] == null ||
      req.variables['APPWRITE_FUNCTION_API_KEY'] == null) {
    print(
        "Environment variables are not set. Function cannot use Appwrite SDK.");
  } else {
    client
        .setEndpoint(req.variables['APPWRITE_FUNCTION_ENDPOINT'])
        .setProject(req.variables['APPWRITE_FUNCTION_PROJECT_ID'])
        .setKey(req.variables['APPWRITE_FUNCTION_API_KEY'])
        .setSelfSigned(status: true);
  }

  res.json({
    'example': req.payload['example'],
  });
}
```
deploying this function and passing the parameter as follows from Appwrite dashboard:
```json
{
  example: "test function"
}
```
it turns out that the state after execution is failed.

Entering the logs in the error section, the following message appears:
An internal curl error has occurred within the executor! Error Message: Operation timed out

We specify that tests have been made with versions prior to 2.17 and also with the node.js runtime

I also want to mention the problem I saw on github and I precise that I followed guide #2940

The issue has also been discussed on the discord channel features

### 👍 Expected behavior

I expect the function to run in a few milliseconds

### 👎 Actual Behavior

![Screenshot 2023-01-27 alle 18 38 23](https://user-images.githubusercontent.com/54111358/215156122-3d74c99a-da6f-4dab-8bb2-5af9620f8688.png)

### 🎲 Appwrite version

Version 1.2.x

### 💻 Operating system

Linux

### 🧱 Your Environment

VM in GCP  => debian-11-bullseye

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

**Comment by stnguyen90** [comment]
@elio1fiore thanks for raising this issue! 🙏 It looks like you're using req.payload as if it's a Map, but it's a String so am exception is probably thrown which leads to the timeout error.

Please make sure your code doesn't throw any exceptions and you can res.json() or res.send() once.

**Comment by elio1fiore** [comment, marked as answer]
Yes the problem is that when you want to use the req.payload you have to use jsonDecode.
In case you are passing the values ​​from the Appwrite console you have to be careful of the recommended json.
<img width="426" alt="Screenshot 2023-01-28 alle 12 47 41" src="https://user-images.githubusercontent.com/54111358/215264808-566125aa-f4e3-420f-b8e0-f0345ed93358.png">
It is not

```json
{
    firstName: "hello", 
    lastName:"world", 
    age:"old"
}
```
but 
```json
{
    "firstName": "hello", 
    "lastName":"world", 
    "age":"old"
}
```

For using req.payload in functions I found no documentation. Since everything was not typed, I had a bit of trouble figuring out what type of data the req.payload was:

```dart
  final payload = jsonDecode(req.payload);

  res.json({'example': payload['example']});

```

**Reply by stnguyen90** [reply]
> In case you are passing the values ​​from the Appwrite console you have to be careful of the recommended json.

Yes, we have an issue open for that: https://github.com/appwrite/appwrite/issues/4940

> Since everything was not typed, I had a bit of trouble figuring out what type of data the req.payload was

Ya, we tried to indicate around our [docs](https://appwrite.io/docs/functions#writingYourOwnFunction) and [templates](https://github.com/appwrite/functions-starter/blob/0bbe60269c7e8212e70ce4a03b9fee3df7f7eee7/dart-2.16/lib/main.dart#L6) that `req.payload` is a `string`. I am pushing to get better typing, though. We'll see how that goes.
